### PR TITLE
Add support for disabling spans and ignoring error log stacktraces.

### PR DESCRIPTION
**Pull request by xzxiong**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #

### Part I: Disable Span
1. action with new configure
	- `disableSpan` bool, default: false. If true, then NOT collect span info.
	- `longSpanTime` toml.Duration, default: 1s. If span's duration < logSpanTime, then NOT record span info.

### Part II: Remove error-level's stacktrace info
1.  new configure
	- `StacktraceLevel` string, default: panic. If log's level < StacktraceLevel, then NOT output stacktrace info.

### Part III: Enhance log output
1.  Ensure ETL log/statement/metric collector quit after all NODE service done.
	- define `serviceWG` in mo-service/main.go to mark all services quit state.
	- ETL collector will quit after all serviceWG Done.

2. Ensure the Fatal log can be saved in MO (make sure log sync before exit)
	- Add interface NeedCheckWrite, to wait signal that item has been wirte down.
	- Add interface NeedSyncWrite, to trigger item to be write immediately
	- Add reactWriter, it can handle NeedCheckWrite hook after write data.

### CODE REVIEW

File: pkg/util/trace/impl/motrace/syncer.go
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/syncer.go:84]: `ItemSyncer` core function, that can receive write-down signal.

File: pkg/util/trace/impl/motrace/mo_trace.go
Comment (xzxiong) [inline pkg/util/trace/impl/motrace/mo_trace.go:149]: new config `longSpanTime` logic

**Comment by matrix-meow** [comment]
@xzxiong Thanks for your contributions!

**Here are review comments for file cmd/mo-service/config.go:**
The pull request adds three main features to the codebase: disabling span, removing error-level's stacktrace info, and enhancing log output. The changes made in the pull request are focused on the `cmd/mo-service/config.go` file. 

There are a few issues with the pull request that need to be addressed:

1. The title of the pull request is not clear and concise. It should be updated to reflect the main changes made in the pull request.

2. The body of the pull request is not well-organized. It should be divided into sections that clearly explain the changes made and why they were necessary. Additionally, the body should include more information about the problem(s) that the pull request is trying to solve.

3. The pull request includes changes to the `cmd/mo-service/config.go` file, but it is not clear how these changes relate to the new features added in the pull request. The pull request should include more information about how these changes are necessary for the new features to work properly.

4. The pull request includes new configurations for disabling span and removing error-level's stacktrace info, but it is not clear how these configurations should be used. The pull request should include more information about how to use these configurations and what their default values are.

5. The pull request includes enhancements to log output, but it is not clear how these enhancements improve the codebase. The pull request should include more information about how these enhancements improve the logging functionality.

To optimize the changes made in the pull request, the following suggestions can be made:

1. The pull request should include more tests to ensure that the new features work properly.

2. The pull request should include more documentation to help developers understand how to use the new features and configurations.

3. The pull request should include more information about how the new features and configurations affect performance.

**Here are review comments for file cmd/mo-service/main.go:**
This pull request adds three main features to the codebase. Firstly, it adds the ability to disable span collection and to ignore span information if the duration is less than a specified time. Secondly, it adds the ability to remove stacktrace information for log levels below a certain threshold. Finally, it enhances log output by ensuring that ETL log/statement/metric collector quits after all NODE services are done, and that Fatal logs can be saved in MO.

One issue with this pull request is that the body is not very clear and concise. It would be helpful to have a brief summary of the changes made in the pull request, as well as a clear explanation of why these changes are necessary. Additionally, the body could benefit from more detailed explanations of the changes made in each part.

Another issue is that the pull request does not include any tests or documentation. It would be helpful to have tests to ensure that the new features are working as intended, and documentation to explain how to use these features.

To optimize the changes made in the pull request, it would be helpful to break the changes down into smaller, more manageable parts. This would make it easier to review and test the changes, and would also make it easier to understand how the changes fit together. Additionally, it would be helpful to include more comments in the code to explain what each part is doing.

Overall, this pull request adds some useful features to the codebase, but could benefit from clearer explanations and more thorough testing and documentation.

**Here are review comments for file pkg/common/log/logger.go:**
The pull request titled "support disable span and ignore error log stacktrace" aims to add new features to the logging functionality of the codebase. The pull request includes three parts: 

Part I: Disable Span - This part adds a new configuration option to disable the collection of span information. The new configuration option is a boolean value named `disableSpan`, which is set to false by default. If set to true, span information will not be collected. Another configuration option named `longSpanTime` is also added, which is a duration value set to 1 second by default. If the duration of a span is less than `longSpanTime`, then span information will not be recorded.

Part II: Remove error-level's stacktrace info - This part adds a new configuration option named `StacktraceLevel`, which is set to "panic" by default. If the log's level is less than `StacktraceLevel`, then stacktrace information will not be output.

Part III: Enhance log output - This part includes two enhancements. First, it ensures that the ETL log/statement/metric collector quits after all NODE services are done. This is achieved by defining a `serviceWG` in `mo-service/main.go` to mark all services' quit state. The ETL collector will quit after all `serviceWG` Done. Second, it ensures that the Fatal log can be saved in MO by adding three new interfaces: `NeedCheckWrite`, `NeedSyncWrite`, and `reactWriter`. `NeedCheckWrite` waits for a signal that an item has been written down, `NeedSyncWrite` triggers an item to be written immediately, and `reactWriter` handles the `NeedCheckWrite` hook after writing data.

The changes made to the `logger.go` file include removing the `AddStacktrace` option from the logger and passing the logger and context to a new `MOLogger` instance.

One potential problem with this pull request is that it includes multiple changes that are not directly related to each other. It may be better to split the pull request into smaller, more focused ones, each addressing a specific issue. Additionally, the pull request could benefit from more detailed explanations of the new features and how they work. Finally, it may be helpful to include examples of how to use the new configuration options in the documentation.

**Here are review comments for file pkg/config/configuration.go:**
The pull request adds three new configurations to the `ObservabilityParameters` struct: `DisableSpan`, `LongSpanTime`, and `StacktraceLevel`. It also adds two new interfaces, `NeedCheckWrite` and `NeedSyncWrite`, and a new struct, `reactWriter`. Additionally, it modifies the `SetDefaultValues` function to set default values for the new configurations.

There are a few issues with this pull request that need to be addressed:

1. The pull request does not provide enough context for why these changes are necessary. It would be helpful to explain why disabling span collection and removing stacktrace info are important, and how the new log output enhancements improve the system.

2. The pull request does not include any tests for the new configurations or interfaces. It is important to ensure that these changes do not introduce any bugs or regressions, so tests should be added to cover the new functionality.

3. The `LongSpanTime` configuration has a default value of 1 second, but the code sets a default value of 500 milliseconds. This inconsistency should be addressed by either updating the default value in the code or updating the documentation to reflect the correct default value.

4. The `LongSpanTime` configuration should have a minimum value of 0, since a negative value does not make sense. The code should be updated to enforce this constraint.

5. The `StacktraceLevel` configuration should have a default value of "error", since it is intended to control whether stacktraces are included in error logs. The code should be updated to reflect this default value.

6. The `NeedCheckWrite` and `NeedSyncWrite` interfaces are not well-documented, and it is not clear how they are intended to be used. It would be helpful to provide more information about these interfaces and how they should be implemented.

7. The `reactWriter` struct is not well-documented, and it is not clear how it is intended to be used. It would be helpful to provide more information about this struct and how it should be used.

8. The pull request modifies the `SetDefaultValues` function, but it does not explain why these changes are necessary or how they impact the system. It would be helpful to provide more information about why these changes were made and how they affect the system.

To optimize the changes made in the pull request, the following suggestions could be considered:

1. Add more detailed documentation to the pull request to explain the new configurations, interfaces, and struct, and how they should be used.

2. Add tests to cover the new functionality, to ensure that it works as intended and does not introduce any bugs or regressions.

3. Consider adding more detailed error messages or logging to help diagnose issues related to the new configurations.

4. Consider adding more detailed comments to the code to help explain how the new functionality works.

**Here are review comments for file pkg/logutil/api.go:**
The pull request titled "support disable span and ignore error log stacktrace" aims to add new features to the logging system. The pull request includes three parts: 

Part I: Disable Span - This feature adds a new configuration option to disable the collection of span information. If the `disableSpan` option is set to true, span information will not be collected. Additionally, a new configuration option `longSpanTime` is added to set the minimum duration for a span to be recorded.

Part II: Remove error-level's stacktrace info - This feature adds a new configuration option to remove stacktrace information from error-level logs. If the log level is less than the `StacktraceLevel` option, stacktrace information will not be output.

Part III: Enhance log output - This feature includes two enhancements to the logging system. First, it ensures that the ETL log/statement/metric collector quits after all NODE services are done. Second, it ensures that the Fatal log can be saved in MO by adding two new interfaces `NeedCheckWrite` and `NeedSyncWrite` and a new `reactWriter` to handle the `NeedCheckWrite` hook after writing data.

The changes made to the `api.go` file remove the `zap.AddStacktrace` option from the logger options.

There are a few issues with this pull request that need to be addressed:

1. Lack of context - The pull request does not provide enough context about the problem it is trying to solve. It would be helpful to include more information about why these features are needed and how they will benefit the system.

2. Incomplete information - The pull request does not include any information about how to use the new configuration options or interfaces. It would be helpful to include examples or documentation on how to use these new features.

3. Lack of testing - The pull request does not include any tests for the new features. It is important to test these new features thoroughly to ensure they work as expected and do not introduce any new bugs.

To optimize the changes made in the pull request, the following suggestions can be made:

1. Provide more context - Include more information about the problem the pull request is trying to solve and how these new features will benefit the system.

2. Include documentation - Include documentation on how to use the new configuration options and interfaces.

3. Add tests - Add tests for the new features to ensure they work as expected and do not introduce any new bugs.

4. Consider refactoring - The pull request includes changes to multiple parts of the system. It may be beneficial to split these changes into separate pull requests to make the review process easier and to ensure that each change is thoroughly tested.

**Here are review comments for file pkg/logutil/dragonboat.go:**
The pull request adds three main features to the codebase:

1. Ability to disable span collection and ignore span info for short spans.
2. Ability to remove stacktrace info for log levels below a certain threshold.
3. Enhancements to log output, including ensuring ETL log/statement/metric collector quits after all NODE service is done and ensuring that Fatal log can be saved in MO.

However, the pull request has some issues that need addressing:

1. The pull request does not reference any issue number, which makes it difficult to track the context of the changes.
2. The pull request does not provide enough information about the changes made to the codebase. The body of the pull request is too brief and does not explain the rationale behind the changes.
3. The changes made to the codebase are not optimized. The pull request adds unnecessary code to the codebase, which can make it harder to maintain and debug in the future. For example, the addition of the `NeedCheckWrite` and `NeedSyncWrite` interfaces and the `reactWriter` can be simplified.
4. The changes made to the codebase are not well-documented. The pull request does not provide enough comments or documentation to help other developers understand the changes made to the codebase.

**Here are review comments for file pkg/logutil/internal.go:**
This pull request adds support for disabling span and ignoring error log stacktrace. It also enhances log output and ensures that the ETL log/statement/metric collector quits after all NODE service is done. Additionally, it ensures that the Fatal log can be saved in MO by adding interfaces NeedCheckWrite and NeedSyncWrite, and a reactWriter to handle NeedCheckWrite hook after writing data.

The changes made in the `pkg/logutil/internal.go` file include adding a new `StacktraceLevel` field to the `LogConfig` struct, which is used to control whether stacktrace information is outputted for logs with a level lower than the specified level. The `initMOLogger` function is modified to accept additional options, and the `getOptions` function is modified to include the `StacktraceLevel` option. The `replaceGlobalLogger` function is modified to remove the `AddStacktrace` option for error logs.

However, there are some issues with this pull request that need to be addressed:

1. Lack of context in the title: The title of the pull request is not descriptive enough and lacks context. It would be helpful to include more information about what this pull request is trying to achieve.

Suggestion: Update the title to include more context, such as "Add support for disabling span and ignoring error log stacktrace".

2. Incomplete documentation: The pull request does not include sufficient documentation for the changes made. It is important to provide clear and concise documentation to help other developers understand the changes made and how to use them.

Suggestion: Add documentation to the pull request, including examples of how to use the new features.

3. Inconsistent naming convention: The naming convention used in the pull request is inconsistent. For example, the `longSpanTime` field is named differently from the `StacktraceLevel` field.

Suggestion: Use a consistent naming convention throughout the pull request.

4. Unnecessary panic: The `getStacktraceLevel` function panics if the `StacktraceLevel` field is not valid. This is unnecessary and can cause issues if the function is called with an invalid value.

Suggestion: Return an error instead of panicking, and handle the error in the calling function.

5. Redundant code: The `replaceGlobalLogger` function is modified to remove the `AddStacktrace` option for error logs, but this option is already included in the `getOptions` function. This is redundant and can cause confusion.

Suggestion: Remove the `AddStacktrace` option from the `getOptions` function, since it is already included in the `replaceGlobalLogger` function.

6. Lack of testing: The pull request does not include any tests for the new features. It is important to include tests to ensure that the new features work as expected and do not introduce any regressions.

Suggestion: Add tests for the new features, including edge cases and error handling.

**Here are review comments for file pkg/logutil/internal_test.go:**
The pull request adds three main features: 

1. Ability to disable span collection and recording based on a new configuration option `disableSpan` and `longSpanTime`.
2. Ability to remove stacktrace information from error-level logs based on a new configuration option `StacktraceLevel`.
3. Enhancements to log output to ensure ETL log/statement/metric collector quits after all NODE service is done and to ensure that Fatal logs are saved in MO.

However, there are a few issues with the pull request that need to be addressed:

1. The title of the pull request is not descriptive enough. It should provide more information about the changes made in the pull request.
2. The body of the pull request is not well-structured. It should be broken down into sections that describe each of the three main features added.
3. The changes made to the `internal_test.go` file are not well-explained. It is not clear why the `StacktraceLevel` option was added or what its default value is.
4. The pull request lacks unit tests to ensure that the new features work as expected.
5. The pull request could benefit from more detailed documentation on how to use the new features.

**Here are review comments for file pkg/util/export/batch_processor.go:**
The pull request adds three new features to the codebase. The first feature is the ability to disable span collection and not record span info if the duration is less than a specified time. The second feature is the ability to remove stacktrace info for error-level logs if the log level is less than a specified level. The third feature is an enhancement to log output to ensure that the ETL log/statement/metric collector quits after all NODE services are done, and the Fatal log can be saved in MO.

The changes made in the pull request are in the file `pkg/util/export/batch_processor.go`. The changes add a new import statement for `table`, and a new condition to signal the buffer holder to flush if the item implements the `NeedSyncWrite` interface.

There are a few issues with the pull request that need addressing:

1. The title of the pull request is not descriptive enough. It should provide a clear and concise summary of the changes made in the pull request.

2. The body of the pull request is missing some important information. It should provide a detailed explanation of the changes made in the pull request, including how the new features work and why they are necessary.

3. The changes made in the pull request are not optimized. The new condition added to signal the buffer holder to flush can be combined with the existing condition to reduce code duplication.

To address these issues, the following suggestions can be made:

1. The title of the pull request can be changed to "Add support for disabling span collection and removing stacktrace info for error-level logs."

2. The body of the pull request can be updated to provide a more detailed explanation of the new features. For example, it can explain how the `disableSpan` and `longSpanTime` options work, and why they are useful. It can also explain how the `StacktraceLevel` option works, and why it is important to remove stacktrace info for error-level logs. Additionally, it can provide more information about the enhancements to log output, including how the `NeedCheckWrite` and `NeedSyncWrite` interfaces work, and why they are necessary.

3. The changes made in the pull request can be optimized by combining the new condition with the existing condition. For example, the following code can replace the changes made in the pull request:

```
if buf.ShouldFlush() || (checker, is := item.(table.NeedSyncWrite); is && checker.NeedSyncWrite()) {
    b.signal(b)
}
```

**Here are review comments for file pkg/util/export/table/type.go:**
The title of the pull request is clear and concise. It indicates that the changes made in the pull request are meant to support disabling span and ignoring error log stacktrace.

The body of the pull request is well-structured and provides a clear explanation of the changes made. However, it could be improved by providing more context about the problem that these changes are meant to solve. Additionally, the body should include information about how the changes were tested and any potential side effects that may arise from these changes.

The changes made in the pull request include adding new configurations to disable span and remove error-level's stacktrace info. The pull request also enhances log output by ensuring that ETL log/statement/metric collector quits after all NODE service is done and that the Fatal log can be saved in MO. The changes made to the file pkg/util/export/table/type.go include adding new interfaces to cooperate with RowWriter and AfterWrite.

One potential problem with the changes is that they may introduce new bugs or issues that were not present before. To address this, the changes should be thoroughly tested before merging into the codebase. Additionally, the pull request should include information about how the changes were tested and any potential side effects that may arise from these changes.

Another issue with the pull request is that it does not provide enough context about the problem that these changes are meant to solve. To address this, the pull request should include information about the problem that these changes are meant to solve and how the changes will address the problem.

Overall, the changes made in the pull request seem reasonable and well-structured. However, the pull request could be improved by providing more context about the problem that these changes are meant to solve and how the changes were tested. Additionally, the pull request should include information about any potential side effects that may arise from these changes.

**Here are review comments for file pkg/util/export/writer_factory.go:**
The pull request titled "support disable span and ignore error log stacktrace" aims to add new features and enhance the logging output. The changes include:

1. Adding a new configuration option to disable span collection and set a threshold for the span duration.
2. Adding a new configuration option to remove stacktrace information for logs below a certain level.
3. Enhancing the log output by ensuring the ETL log/statement/metric collector quits after all NODE services are done and ensuring that the Fatal log can be saved in MO.

The changes made to the `writer_factory.go` file include adding a new `reactWriter` struct that implements the `table.AfterWrite` interface and can react before/after `FlushAndClose`. The `GetWriterFactory` function has been modified to use the `newWriter` function to create a new `reactWriter` instance.

There are a few issues with this pull request that need addressing:

1. The pull request title is not descriptive enough. A more descriptive title would help reviewers understand the purpose of the pull request at a glance.
2. The body of the pull request is not well-organized. It would be helpful to separate the changes into different sections and provide more context for each change.
3. The changes made to the `writer_factory.go` file are not well-explained. It would be helpful to provide more context for why the `reactWriter` struct was added and how it improves the code.
4. The code changes could benefit from more comments and documentation to help future maintainers understand the code.

To address these issues, the pull request author could consider the following suggestions:

1. Update the pull request title to be more descriptive, such as "Add configuration options for disabling span collection and removing stacktrace information".
2. Organize the body of the pull request into separate sections for each change and provide more context for each change.
3. Provide more explanation for the changes made to the `writer_factory.go` file, including how the `reactWriter` struct improves the code and why it was added.
4. Add more comments and documentation to the code changes to help future maintainers understand the code.

**Here are review comments for file pkg/util/trace/impl/motrace/buffer_pipe.go:**
The pull request adds three main features to the codebase:

1. Ability to disable span collection and ignore span info if the duration is less than a specified time.
2. Ability to remove stacktrace info for log levels below a specified level.
3. Enhancements to log output to ensure ETL log/statement/metric collector quits after all NODE service is done and to ensure that the Fatal log can be saved in MO.

However, the pull request lacks some important information that would help reviewers understand the changes better. For example:

1. The pull request does not explain how the new features work or how to use them.
2. The pull request does not provide any examples or use cases for the new features.
3. The pull request does not explain why the changes were made or what problem they solve.

To address these issues, the pull request should include more detailed explanations of the new features, including how to use them and why they were added. Additionally, the pull request should include examples or use cases to help reviewers understand the changes better. Finally, the pull request should be optimized by removing any unnecessary changes or code that does not contribute to the new features.

**Here are review comments for file pkg/util/trace/impl/motrace/buffer_pipe_test.go:**
The pull request adds three main features: disabling span, removing error-level's stacktrace info, and enhancing log output. The changes made in the pull request are in the file `pkg/util/trace/impl/motrace/buffer_pipe_test.go`. 

There are a few issues with this pull request that need to be addressed:

1. The title of the pull request is not descriptive enough. It should provide a clear and concise summary of the changes made in the pull request.

2. The body of the pull request is missing some important information. It should include details on how to use the new features and how they will benefit the codebase.

3. The changes made in the pull request are not optimized. The changes should be split into separate pull requests to make it easier to review and merge.

4. The changes made in the file `pkg/util/trace/impl/motrace/buffer_pipe_test.go` are not explained in the pull request. It is unclear why these changes were made and how they relate to the new features.

To address these issues, the pull request should be updated with the following changes:

1. The title of the pull request should be updated to provide a clear and concise summary of the changes made. For example, "Add support for disabling span and removing error-level's stacktrace info".

2. The body of the pull request should be updated to include details on how to use the new features and how they will benefit the codebase. For example, it should include examples of how to use the new configuration options and how they will improve performance.

3. The changes made in the pull request should be split into separate pull requests to make it easier to review and merge. For example, one pull request could focus on disabling span, another on removing error-level's stacktrace info, and another on enhancing log output.

4. The changes made in the file `pkg/util/trace/impl/motrace/buffer_pipe_test.go` should be explained in the pull request. It should be clear why these changes were made and how they relate to the new features. For example, the changes could be related to testing the new configuration options.

**Here are review comments for file pkg/util/trace/impl/motrace/config.go:**
The pull request proposes several changes to the codebase, including the addition of new configurations and enhancements to log output. However, there are a few issues that need to be addressed:

1. The title of the pull request is not descriptive enough. It should provide a clear and concise summary of the changes made in the pull request.

2. The body of the pull request is not well-organized. It should be divided into sections that clearly explain the changes made and why they were necessary. Additionally, the body should include information on how to test the changes.

3. The changes made to the `config.go` file are not well-documented. Each configuration option should have a clear description of what it does and how it should be used.

4. The pull request introduces several new interfaces (`NeedCheckWrite`, `NeedSyncWrite`, and `reactWriter`) without providing sufficient explanation of their purpose and how they should be used.

To address these issues, the pull request should be updated as follows:

1. The title of the pull request should be updated to provide a clear and concise summary of the changes made. For example, "Add support for disabling spans and ignoring error log stacktraces."

2. The body of the pull request should be divided into sections that clearly explain the changes made and why they were necessary. Each section should include information on how to test the changes. For example:

   - Introduction: Provide an overview of the changes made in the pull request.
   - Part I: Disable Span: Explain the new `disableSpan` configuration option and how it can be used to disable span collection. Provide an example of how to use this option.
   - Part II: Remove error-level's stacktrace info: Explain the new `StacktraceLevel` configuration option and how it can be used to control when stacktraces are included in error logs. Provide an example of how to use this option.
   - Part III: Enhance log output: Explain the new interfaces (`NeedCheckWrite`, `NeedSyncWrite`, and `reactWriter`) and how they can be used to improve log output. Provide an example of how to use these interfaces.

3. The changes made to the `config.go` file should be well-documented. Each configuration option should have a clear description of what it does and how it should be used. For example:

   - `disableSpan`: If true, disables span collection.
   - `longSpanTime`: If a span's duration is less than this value, span information will not be recorded.
   - `StacktraceLevel`: Determines the log level at which stacktraces will be included in error logs.
   - `NeedCheckWrite`: Interface that waits for a signal indicating that an item has been written.
   - `NeedSyncWrite`: Interface that triggers an item to be written immediately.
   - `reactWriter`: Handles `NeedCheckWrite` hook after writing data.

4. The pull request should provide more information on the purpose and usage of the new interfaces (`NeedCheckWrite`, `NeedSyncWrite`, and `reactWriter`). This could include code examples and explanations of how these interfaces improve log output.

**Here are review comments for file pkg/util/trace/impl/motrace/mo_trace.go:**
The pull request titled "support disable span and ignore error log stacktrace" includes changes to the `mo_trace.go` file in the `pkg/util/trace/impl/motrace` directory. The changes include three parts:

Part I: Disable Span
- Adds a new configuration option `disableSpan` that, when set to `true`, disables the collection of span information.
- Adds a new configuration option `longSpanTime` that, when set to a duration, determines the minimum duration of a span that will be recorded.

Part II: Remove error-level's stacktrace info
- Adds a new configuration option `StacktraceLevel` that, when set to a log level, determines the minimum log level at which stacktrace information will be output.

Part III: Enhance log output
- Ensures that ETL log/statement/metric collector quits after all NODE services are done.
- Ensures that the Fatal log can be saved in MO by adding two new interfaces, `NeedCheckWrite` and `NeedSyncWrite`, and a new `reactWriter` that handles the `NeedCheckWrite` hook after writing data.

The changes to the `mo_trace.go` file include modifying the `MOSpan` struct to change the `Duration` field from a `uint64` to a `time.Duration`. The `End` method is also modified to set the `EndTime` and `Duration` fields and to check if the span's duration is less than the `longSpanTime` configuration option before recording the span.

There are a few issues with this pull request that need to be addressed:

1. Lack of context in the body: The body of the pull request does not provide enough context about why these changes are being made or what problem they are solving. It would be helpful to include more information about the motivation behind these changes.

2. Inconsistent naming convention: The `longSpanTime` configuration option uses camel case, while the other configuration options use snake case. It would be better to use a consistent naming convention throughout the codebase.

3. Confusing comment: The comment for the `Duration` field in the `MOSpan` struct is unclear. It would be better to provide a more detailed explanation of what this field represents.

4. Redundant code: The `if s.EndTime.IsZero()` check in the `End` method is redundant since the `EndTime` field is set in the previous line.

To optimize the changes made in the pull request, it would be helpful to provide more context in the body of the pull request and to use a consistent naming convention for configuration options. Additionally, the comment for the `Duration` field in the `MOSpan` struct should be clarified, and the redundant code in the `End` method should be removed.

**Here are review comments for file pkg/util/trace/impl/motrace/report_log.go:**
The pull request adds support for disabling span and ignoring error log stacktrace. It also enhances log output by ensuring ETL log/statement/metric collector quits after all NODE service is done and ensuring the Fatal log can be saved in MO. 

There are a few issues with the pull request that need addressing:

1. The title of the pull request is not descriptive enough. It should provide more information about the changes made in the pull request.

2. The body of the pull request is not well-organized. It should be broken down into sections that clearly explain the changes made and why they are necessary. 

3. The changes made to the file `pkg/util/trace/impl/motrace/report_log.go` are not optimized. The switch statement added to handle different log levels can be simplified by moving the common code outside the switch statement. 

To address these issues, the following suggestions can be implemented:

1. The pull request title should be updated to provide more information about the changes made. For example, "Add support for disabling span and ignoring error log stacktrace."

2. The body of the pull request should be organized into sections that clearly explain the changes made and why they are necessary. For example:

   - Introduction: Briefly explain the purpose of the pull request.
   - Part I: Disable Span: Explain the new configuration options added to disable span and the conditions under which span info is not collected.
   - Part II: Remove error-level's stacktrace info: Explain the new configuration option added to remove error-level's stacktrace info.
   - Part III: Enhance log output: Explain the changes made to ensure ETL log/statement/metric collector quits after all NODE service is done and to ensure the Fatal log can be saved in MO.
   
3. The changes made to the file `pkg/util/trace/impl/motrace/report_log.go` can be optimized by moving the common code outside the switch statement. For example:

```
func ReportZap(jsonEncoder zapcore.Encoder, entry zapcore.Entry, fields []zapcore.Field) (*buffer.Buffer, error) {
	buffer, err := jsonEncoder.EncodeEntry(entry, fields)
	if err != nil {
		return nil, err
	}
	log := &Log{
		Timestamp: time.Now().UnixNano(),
		Level:     int32(entry.Level),
		Message:   buffer.String(),
	}
	if entry.Caller != nil {
		log.Caller = entry.Caller.String()
	}
	if len(fields) > 0 {
		log.Fields = *(*[]Field)(unsafe.Pointer(&fields))
	}
	switch entry.Level {
	case zap.PanicLevel, zap.DPanicLevel, zap.FatalLevel:
		syncer := NewItemSyncer(log)
		GetGlobalBatchProcessor().Collect(DefaultContext(), syncer)
		syncer.Wait()
	default:
		GetGlobalBatchProcessor().Collect(DefaultContext(), log)
	}
	return buffer, nil
}
```

By making these changes, the pull request will be more organized and optimized, making it easier to review and merge.

**Here are review comments for file pkg/util/trace/impl/motrace/syncer.go:**
The pull request titled "support disable span and ignore error log stacktrace" aims to add new features and enhancements to the codebase. The pull request includes three parts: 

1. Part I: Disable Span - adds a new configuration option to disable span collection and another option to ignore spans with a duration less than a specified value.
2. Part II: Remove error-level's stacktrace info - adds a new configuration option to remove stacktrace information from logs with a level lower than a specified value.
3. Part III: Enhance log output - adds two new interfaces to handle log writing and synchronization.

The changes made in the pull request are limited to the file `pkg/util/trace/impl/motrace/syncer.go`. The changes add new interfaces and methods to the `ItemSyncer` struct.

While the pull request seems to add useful features, there are several issues that need addressing:

1. Lack of context - The pull request lacks context about the problem it aims to solve. It would be helpful to have more information about why these features are necessary and how they will benefit the codebase.

2. Incomplete documentation - The pull request lacks documentation for the new features and interfaces added. It would be helpful to have clear documentation on how to use the new features and interfaces.

3. Inconsistent naming - The naming of the new interfaces and methods is inconsistent with the existing codebase. It would be helpful to follow the naming conventions used in the codebase to maintain consistency.

4. Lack of testing - The pull request lacks tests for the new features and interfaces added. It would be helpful to have tests to ensure that the new features and interfaces work as intended.

To optimize the changes made in the pull request, the following suggestions can be considered:

1. Provide more context - Add more information about the problem the pull request aims to solve and how the new features will benefit the codebase.

2. Improve documentation - Add clear documentation on how to use the new features and interfaces.

3. Follow naming conventions - Follow the naming conventions used in the codebase to maintain consistency.

4. Add tests - Add tests to ensure that the new features and interfaces work as intended.

**Here are review comments for file pkg/util/trace/impl/motrace/trace.go:**
The pull request adds support for disabling span and ignoring error log stacktrace. It also enhances log output by ensuring ETL log/statement/metric collector quits after all NODE service is done and ensuring the Fatal log can be saved in MO. 

There are a few issues with this pull request that need addressing:

1. The title of the pull request is not descriptive enough. It should provide more information about the changes made in the pull request.

2. The body of the pull request is missing some important information. It should provide more context about why these changes are necessary and what problems they solve.

3. The changes made in the pull request are not optimized. For example, the addition of `WithLongSpanTime` and `WithSpanDisable` could have been combined into a single function to reduce code duplication.

4. The code changes are not well-documented. It would be helpful to add comments to explain what each function does and why it is necessary.

5. The pull request does not include any tests. It is important to ensure that the changes made do not introduce any new bugs or regressions.

To address these issues, the pull request should be updated with a more descriptive title and a more detailed body that explains the motivation behind the changes. The changes should also be optimized to reduce code duplication, and the code should be well-documented with comments. Finally, tests should be added to ensure that the changes do not introduce any new bugs or regressions.

**Here are review comments for file test/distributed/cases/log/query_tae_table.result:**
The pull request adds several new features and enhancements to the codebase. However, there are several issues that need to be addressed:

1. The title of the pull request is not descriptive enough. It should provide a clear and concise summary of the changes made in the pull request.

2. The body of the pull request is too long and contains unnecessary information. It should be shortened and focus on the key changes made in the pull request.

3. The pull request includes changes to a test file, which is not recommended. Test files should be kept separate from code changes to avoid confusion and maintain clarity.

4. The pull request introduces several new configurations and interfaces, which may make the code more complex and difficult to maintain. It would be helpful to provide more information on why these changes are necessary and how they will benefit the codebase.

5. The pull request includes changes to log output, which may impact the performance of the application. It would be helpful to provide more information on how these changes were tested and what impact they may have on the application.

To optimize the changes made in the pull request, it would be helpful to:

1. Provide more detailed explanations of the changes made in the pull request, including why they were necessary and how they will benefit the codebase.

2. Separate the changes made to the test file into a separate pull request to maintain clarity and avoid confusion.

3. Simplify the new configurations and interfaces to make the code more maintainable and easier to understand.

4. Provide more information on how the changes to log output were tested and what impact they may have on the application.

**Here are review comments for file test/distributed/cases/log/query_tae_table.sql:**
The pull request adds several improvements to the codebase, including the ability to disable span, remove error-level's stacktrace info, and enhance log output. However, there are several issues that need addressing:

1. The pull request title is not descriptive enough. It should provide a clear and concise summary of the changes made in the pull request.

2. The body of the pull request is too long and contains unnecessary information. It should focus on providing a brief overview of the changes made and why they are necessary.

3. The pull request does not reference any specific issues that it fixes. It should reference the relevant issue(s) in the body of the pull request.

4. The changes made to the test/distributed/cases/log/query_tae_table.sql file are not explained in the body of the pull request. It is unclear why these changes were made and how they relate to the other changes in the pull request.

To address these issues, the pull request title should be updated to provide a clear and concise summary of the changes made. The body of the pull request should be shortened and focus on providing a brief overview of the changes made and why they are necessary. The pull request should reference the relevant issue(s) that it fixes, and the changes made to the test/distributed/cases/log/query_tae_table.sql file should be explained in the body of the pull request.

**Review by fengttt — APPROVED** [review]
StackTraceLevel -- This is tricky.    Most likely in prod env we set this at panic level.    However, I do want to see stack of those "internal errors", because internal errors should never reach user, and in fact every internal error happened is a bug, so we want to see stack and fix.

That said, it probably is OK -- for internal errors, which are bugs, maybe we should always panic ...

**Review by daviszhen — APPROVED** [review]

**Review by heni02 — APPROVED** [review]

**Review by zhangxu19830126 — APPROVED** [review]

**Review by gavinyue — APPROVED** [review]

**Pull request merged**
